---
title: "QTL2"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This code allows for the creation of QTL scans through QTL2 and some rudimentary exploration under the peaks found on that scan.

```{r}
library(qtl2scan)
library(qtl2plot)
library(qtl2geno)
library(qtl2convert)
```

Step one is to load in all ofthe packages. 

```{r, eval = FALSE}
oldpheno <- read.csv('Churchill_project/Data/data_restructured.csv',header=T)
newpheno <- oldpheno[match(names(probs.478[,1,1]), oldpheno$Sample),]
rownames(newpheno) <- newpheno[,3]
```

We then load in the data.  In this case, our phenotype data had more mice than the expression data, so we had to trim it down.  Additionally, QTL2 requires the sample IDs be the row names, so we pulled those out as well.

```{r, eval = FALSE}
newpheno$urine.microalbumin1 <- as.numeric(as.character(newpheno$urine.microalbumin1))

rz.transform<-function(y) {
  rankY=rank(y, ties.method="average", na.last="keep")
  rzT=qnorm(rankY/(length(na.exclude(rankY))+1))  
  rzT
}

newpheno.rz <- newpheno
for(i in 9:ncol(newpheno)){
  newpheno.rz[,i] <- rz.transform(newpheno.rz[,i])
}
```

We had to do some reshaping of our phenotype data as well.  Firstly, one of our phenotypes for whatever reason wasn't numeric, so we changed it so that it could be analyzed.  Additionally, we decided to rz transform our data, so that it gave us more normal QTL scans, as un-transformed data soemtimes gives massive peaks.

```{r, eval = FALSE}
covar <- cbind(model.matrix(newpheno[,"Sex"], newpheno[,"Diet"])
colnames(covar)[1:2] = c("Sex", "Diet")
covar.gen <- cbind(covar, newpheno[,5:6])
covar.gen[,3] <- as.numeric(covar.gen[,3])-1
covar.gen[,4] <- as.numeric(covar.gen[,4])-1
colnames(covar.gen)[1:4] = c("Sex", "Diet", "Gen", "parity")

sex <- covar.gen[,1,drop = FALSE]
diet <- covar[,2,drop = FALSE]
```

We then have to create a data frame for our covariates.  We simply extract the information from our phenotype dataset and cbind it together.  It is suggested that we use gen and parity as covariates, as some effects have been seen.  We also make single factor data frames for sex and diet, to use with interactive scans.  While not necessary, they are pretty convenient.

```{r, eval = FALSE}
if (class(probs.478) != "calc_genoprob"){
  probs=probs_doqtl_to_qtl2(probs.478, map=snps.64K,
                            chr_column = "Chr", pos_column = "cM",
                            marker_column = "SNP_ID")
}

qtl2apr = genoprob_to_alleleprob(probs)
```

Our expression data had the genotype probabilities in DOQTL form, so we had to convert them to QTL2 form.  Luckily, QTL2 lets us do this pretty easily.  

```{r, eval = FALSE}
k <- calc_kinship(qtl2apr, "loco", cores=8)
```

We have to calculate the mouse kinships for the qtl scans, and QTL2 makes it pretty easy.  You just have to input the probs and the method you want to use.

```{r, eval = FALSE}
newpheno.rz <- newpheno.rz[,9:ncol(newpheno.rz)]
pp.expr <- expr.rna.478[,"ENSMUSG00000055850", drop = FALSE]
pp=newpheno.rz[,c("BW.10", "BW.12"), drop = FALSE]
```

Some of the data in the phenotype data are factors and therefore non-numeric, so we just cut them out.  Then, to do the QTL scan, we need to create an object that contains just the data of the phenotype or phenotypes we want to scan for. We can use QTL2 to scan multiple phenotypes at the same time simply by adding multiple phenotype to pp, and this is also much faster than doing individual scans for each.

```{r, eval = FALSE}
qtl <- scan1(qtl2apr, pp, k, addcovar = covar.gen[,2:4], intcovar = sex, cores = 4)
```

We can now perform a QTL scan.  As you can see, we just input all of what we've created so far, choosing our interactive covariates from the single coumn data frames we created earlier.

```{r, eval = FALSE}
peak <- bayes_int(allscan1.difdiet[,"BW.4"], chr = NULL)

geneRange <- data.frame()
for(i in 1:nrow(annotations.rna.478)){
  if(annotations.rna.478[i,4] < peak[,3] && annotations.rna.478[i,5] > peak[,1] && annotations.rna.478[i,3] == 10){
    insert <- cbind(annotations.rna.478[i,2], annotations.rna.478[i,4], annotations.rna.478[i,5], i)
    geneRange <- rbind(geneRange, insert)
  }
}
```

We can now take the highest peak on the QTL scan and look at the genes under it by looking through the gene annotation data.  Bayes_int gives us a range within which the gene causing the peak is predicted to be in, and by looking at each gene's position we can find all gene's under the peak.

```{r, eval = FALSE}
for (i in 1:dim(geneRange)[1]){
  covar.drop <- cbind(covar.gen, expr.rna.478[,annotations.rna.478
                                                    [grep(geneRange[i,1], annotations.rna.478[,2]),1]])
  qtl.drop <- scan1(qtl2apr, pp, k, 
                      addcovar = covar.drop[,c(1, 3:5)], intcovar = diet, cores = 7)
  peak.value <- find_peaks(qtl.drop)
  if(peak.value[17,5] < 10.661857){
    print(geneRange[i,1])
  }
}
```

Finally, for each gene found under the peak, we can add that gene as a covariate and find if the QTL peak drops, suggesting that that gene has some effect on the phenotype.
