---
title: "Med_scans"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

```{r, include = FALSE}
load("~/Desktop/Rstudio/Churchill_project/Data/DO478_ExprData_4eQTLMapping.Rdata")
load("~/Desktop/Rstudio/covar.gen.Rdata")
```

```{r, message=FALSE}
devtools::install_github("simecek/intermediate")
library(intermediate)
library(qtl2scan)
library(qtl2plot)
library(qtl2geno)
library(qtl2convert)
```

This code allows for quick creation of interactive mediation scans.  You simply need to fill in the fields below:

```{r, message=FALSE}
pheno.data <- read.csv('~/Desktop/Rstudio/Churchill_project/Data/svenson.phenotypes.final.850.csv',header=T)
phenotype <- "BW.20"
chr <- 9
mediator <- expr.rna.478
annotations <- annotations.rna.478
map <- snps.64K
probs <- probs.478
covar <- covar.gen
addcovar <- c("Diet", "Gen", "parity")
intcovar <- "Sex"
```

* pheno.data is assigned to the phenotype data.
* phenotype is a string that is the name of a phenotype in pheno.data.  This is case sensitive.
* chr should have assigned to it which chromosome you want to look at.  If you simply want to look at the highest peak for the phenotype, chr should be assigned to NULL.
* mediator should be assigned to what the phenotype will be conditioned with, usually the gene exrpession data.
* annotations should have assigned to it the gene annotation information.
* map should have assigned ot it the snps data for the mice.
* probs should have assigned to it the genotype probablity for each mouse.
* covar should have assigned to it a data frame with all covariate information for each mouse that you want to include.
* addcovar should have assigned to it a list of phenotypes, all of which are in covar, that you want to be additive covariates.
* intcovar should be a single string that is the interactive covariate, which should be contained in covar.

```{r, message=FALSE}
phenotype <- as.data.frame(pheno.data[,phenotype])

if("Sample" %in% colnames(pheno.data)){
  rownames(phenotype) <- pheno.data$Sample
}

phenotype <- as.data.frame(phenotype[which(rownames(phenotype) %in% rownames(mediator)),,drop = FALSE])

rz.transform<-function(y) {
  rankY=rank(y, ties.method="average", na.last="keep")
  rzT=qnorm(rankY/(length(na.exclude(rankY))+1))  
  rzT
}

phenotype[,1] <- as.numeric(phenotype[,1])

phenotype[,1] <- rz.transform(phenotype[,1])

phenotype <- phenotype[rownames(expr.rna.478),,drop = FALSE]
```

First we have to extract the phenotype data.  We extract the column with the phenotype from pheno.data.  We then extract the Mouse ID information from pheno.data and label each mouse based on that.  Again, our phenotype data had more mice than the expression data, so we had to cut it down.  Finally, we rankZ transformed the data and reorder it to the order of the expression data.

```{r, message=FALSE}
if (class(probs.478) != "calc_genoprob"){
  probs=probs_doqtl_to_qtl2(probs, map=map,
                            chr_column = "Chr", pos_column = "cM",
                            marker_column = "SNP_ID")
}

qtl2apr = genoprob_to_alleleprob(probs)
```

Our probs data was in DOQTL form, so we had to convert it to QTL2 form.  Luckily QTL2 allows us to easily do this.

```{r, message=FALSE}
k <- calc_kinship(qtl2apr, "loco", cores=8)
```

We have to calculate kinship for the mediation scans, which is quite easy with QTL2.

```{r, message=FALSE}
qtl <- scan1(qtl2apr, phenotype, k, addcovar = covar[,addcovar], intcovar = covar[,intcovar, drop = FALSE])
```

We now have to perform a QTL scan to find the specific peak that we want to look at in the mediation scans.

```{r, message=FALSE}
qtl.geno <- probs.478[,,rownames(max_scan1(qtl, chr = chr))[1]]

if(!is.null(intcovar)){
  qtl.geno <- cbind(qtl.geno, qtl.geno*covar[,intcovar])
}
```

Mediation scans need the genotype probability for the peak that analysis is being done on, so we pull that out of the probs data.  Then, if there is an interactive covariate, we also add an interactive term for each possible genotype by multiplying the genotype probabilities by the interactive covariate.

```{r, message=FALSE}
covar <- covar[,c(addcovar, intcovar)]

for(i in 1:ncol(covar)){
  covar[,i] <- as.numeric(covar[,i])
}
```

In case more covariates are in the  data that is input, we just cut the dataset down to the covariates provided.

```{r, message=FALSE}
index <- grep("chr", colnames(annotations), ignore.case = TRUE)
colnames(annotations)[index] <- "CHR"
index <- grep("Start.Mbp", colnames(annotations), ignore.case = TRUE)
colnames(annotations)[index] <- "POS"
index <- grep("pos", colnames(annotations), ignore.case = TRUE)
colnames(annotations)[index] <- "POS"
```

Mediation scans need a column in the annotations data called CHR and a column named POS.  In our dataset we replaced the start position with POS, but the other code is also included for other datasets.

```{r, message=FALSE, results="hide"}
med.add <- mediation.scan(target = phenotype[,1], mediator = mediator,
                          annotation = annotations, 
                          covar = as.matrix(covar), 
                          qtl.geno = qtl.geno)
```

Finally we can perform the mediation scans, inputting all of the data that we have cleaned in this code.

```{r, results="hide"}
kplot(med.add)
```

This creates an interactive plot of the mediation scan, in which you can hover over any point to find what gene it represents.
